### PR TITLE
Add `Enumerable(T)#to_a(& : T -> U) forall U`

### LINKED ISSUE #12643: Add `Enumerable(T)#to_a(& : T -> U) forall U`
## Feature Request

**Is your feature request related to a problem? Please describe clearly and concisely what is it.**

It's common to find patterns like this in Crystal codebases

```crystal
# Cases where the `Enumerable` implements a covariant `map`
enumerable.map(&.some_attribute).to_a
# Or...
enumerable.to_a.map(&.some_attribute)
```

The `map` would be a redundant iteration if `to_a` could accept a block transform.

The implementation of a `to_a` that accepts a block has been previously [discussed as a way to get around covariance issues](https://github.com/crystal-lang/crystal/issues/9184#issuecomment-619957462).

**Describe the feature you would like, optionally illustrated by examples, and how it will solve the above problem.**

Add `Enumerable(T)#to_a(& : T -> U) forall U`

**Describe considered alternative solutions, and the reasons why you have not proposed them as a solution here.**

Writing a loop yourself to improve performance.

```crystal
array = [] of U
enumerable.each do |item|
  array << item.some_attribute
end
``` 

**Does it break backward compatibility, if yes then what's the migration path?**

Nope. However, it would be simple to create a correctable [ameba](https://github.com/crystal-ameba/ameba) rule that would pick this pattern up and transform it.

### PR DESCRIPTION
**Pull request by caspiano**
Analogous to `Enumerable(T)#(& : T -> Tuple(K, V)) forall K, V` except producing an Array from the block's result.

Resolves #12643

### CODE REVIEW

File: src/enumerable.cr [hunk 1]
```diff
@@ -1804,8 +1804,17 @@ module Enumerable(T)
   # (1..5).to_a # => [1, 2, 3, 4, 5]
   # ```
   def to_a
-    ary = [] of T
-    each { |e| ary << e }
+    to_a(&.itself)
+  end
+
+  # Returns an `Array` with all the elements returned from the *block*.
```
Comment (HertzDevil) [inline src/enumerable.cr:1810]: The documentation should be closer to that of `#map` while highlighting their differences (and vice-versa in `#map`'s documentation too).

Comment (asterite) [reply src/enumerable.cr:1810]: Exactly. I would say that `map` is sometimes overwritten by some types (for example Tuple, Slice) to return the same type, and that if an `Array` is always needed then to use `to_a`.

**Comment by HertzDevil** [comment]
`Indexable` could override this to pre-allocate the returned array with capacity same as `size`. (This does not improve performance in all cases but is usually a right thing to do, like array literals.)

**Comment by asterite** [comment]
> Indexable could override this to pre-allocate the returned array with capacity same as size.

Strange, that's what Array already does. I'm surprised when Indexable was introduced this optimization wasn't also done.
 
> This does not improve performance in all cases

Why not?

**Comment by HertzDevil** [comment]
If you immediately insert an element to the returned array then a reallocation is guaranteed to happen and that is less performant than the default reallocation strategy on average. This is why we didn't pursue #10075

File: src/enumerable.cr [hunk 2]
```diff
@@ -1804,8 +1804,17 @@ module Enumerable(T)
...
+  #
+  # ```
+  # (1..5).to_a { |i| i * 2 } # => [2, 4, 6, 8, 10]
+  # ```
+  def to_a(& : T -> U) forall U
```
Comment (HertzDevil) [inline src/enumerable.cr:1815]: Maybe add return value restrictions?

**Review by straight-shoota — APPROVED** [review]

File: src/hash.cr [hunk 3]
```diff
@@ -1928,16 +1928,28 @@ class Hash(K, V)
...
+  end
+
+  # Returns an `Array` with the results of running *block* against tuples with key and values
+  # belonging to this Hash.
+  #
+  # ```
+  # h = {"first_name" => "foo", "last_name" => "bar"}
+  # h.to_a { |_k, v| v.capitalize } # => ["Foo", "Bar"]
+  # ```
+  # The order of the array follows the order the keys were inserted in the Hash.
```
Comment (HertzDevil) [inline src/hash.cr:1949]:
```suggestion
  # ```
  #
  # The order of the array follows the order the keys were inserted in the Hash.
```

ditto for the original blockless `#to_a` overload

**Review by HertzDevil — CHANGES_REQUESTED** [review]

**Review by HertzDevil — APPROVED** [review]

**Pull request merged**